### ISSUE TITLE
Pansharpening and equivariance

### ISSUE THREAD

**Issue by celoyd** [celoyd/potato#22]
Some slightly inchoate thoughts on equivariance, mostly not ready to be implemented in any particular way. I’m embarrassed to say I know no group theory whatsoever, so this is going to be awkwardly and perhaps inaccurately phrased.

Ideal pansharpening in the continuous world would be equivariant under a wide range of operations. In other words, we expect that:

$$
\psi(p, m) = \mathrm{T^{-1}}\psi(\mathrm{T}(p), \mathrm{T}(m))
$$

where T is one of a large set of transformations that we can think of as roughly coinciding with conformal maps (but see appendix), and T⁻¹ is its inverse. 

In our practical domain of discrete pixel grids, many conformal maps would involve information-destroying resampling. But there are a few specific ones whose equivariances might be realistic to enforce:

* **shifts** (a.k.a. translations) by integer numbers of pixels,
* **rotations** by multiples of 90°, and
* **flips** (a.k.a. reflections).

Unless I’m forgetting something, these are the transforms that are both conformal and non-resampling.

What does this mean? For example, I think it’s reasonable to have a rotation equivariance loss that looks something like:

```python
eq_loss = torch.mean(torch.abs(
    rotate(model(p, m))
    - model(rotate(p), rotate(m))
))
```

where `rotate` is a choice (presumably made at random just above this snippet) of rotating 90°, 180°, or 270°.

## Why, though?

There are two problems I want to address this way, one theoretical and the other practical:

### Asymmetries in the convolutions

Shift equivariance is kind of _the_ big idea of CNNs, but rotation invariance is perhaps underappreciated. Orientation does matter in many vision tasks but I suspect that a lot of low-level operations would be well served by convolutions with weakly enforced symmetries. (The goal isn’t to have literally symmetrical kernels, only to apply a regularizing force away from _unnecessary net asymmetry_ within any substantial part of the network.) Fixed decomposition with wavelets, Gabor filters, etc., is one way to get this; there are others. Mind you, there is some small amount of information in the orientation of satellite imagery (for example, in deciding whether a surface is shadowed or merely dark), but for this deliberately lean project we’re willing to ignore it, and are happy to train the model with rotation-augmented imagery. Overall, we consider any energy the model spends learning an absolute directional dependence to be wasted, and we are interested in discouraging that.

### Griddy output

I’ve noticed that models really like to retain the raw multispectral grid when unsure of how to act. This is of course already under pressure from edge-aware losses etc., but should also be a very strong shift equivariance violation.

## Counterarguments

There’s a reasonable case, I think, that this is not likely to help much.

First because of transitivity: if y is ground truth and ŷ₁ and ŷ₂ are varied outputs, then in principle doing the appropriate loss on the pairs (y, ŷ₁) and (y, ŷ₂) should already be doing anything that (ŷ₁, ŷ₂) can.

Second, (ŷ₁, ŷ₂) has a trivial bad minimum: constant output (e.g., everything is always black).

With these in mind, I think it’s wise to approach equivariant losses as minor tricks that cannot be the dominant losses.

## Shifts have to be on the panchromatic grid

To address griddy output, shifting necessarily has to happen in terms of pan pixels, which interacts with the downsampling (Wald) procedure. This is a headache.

## Appendix: light

Here we’ve only considered geometric equivariance. There is at least one other kind worth thinking about: illumination. In principle, since we’re in the linear space of reflectance, it’s reasonable to say that, for example, it should always be the case that:

$$
\psi(p, m) = \dfrac{\psi(np, nm)}{n}
$$

Where <var>n</var> is a scalar illumination level multiplier. That is, for example, 2× the pixel values of the correct rendering of a scene in half sunlight should be identical to the correct rendering of the same scene in full sunlight. (Per-sample mean normalization would be asserting this implicitly.) As with orientation, there is actually some imaginable information in amount of lighting. As with orientation, it’s not clear that it’s worth worrying about.
